nacos/src/main/java/com/linecorp/armeria/internal/nacos/QueryInstancesClient.java

```diff
@@ -0,0 +1,211 @@
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linecorp.armeria.internal.nacos;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static java.util.Objects.requireNonNull;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.CompletableFuture;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.google.common.base.MoreObjects;
+
+import com.linecorp.armeria.client.Endpoint;
+import com.linecorp.armeria.client.WebClient;
+import com.linecorp.armeria.common.HttpEntity;
+import com.linecorp.armeria.common.QueryParams;
+import com.linecorp.armeria.common.annotation.Nullable;
+
+/**
+ * A Nacos client that is responsible for
+ * <a href="https://nacos.io/en-us/docs/v2/guide/user/open-api.html">Nacos Open-Api - Query instances</a>.
+ */
+final class QueryInstancesClient {
+
+    static QueryInstancesClient of(NacosClient nacosClient, String nacosApiVersion, String serviceName,
+                                   @Nullable String namespaceId, @Nullable String groupName,
+                                   @Nullable String clusterName, @Nullable Boolean healthyOnly,
+                                   @Nullable String app) {
+        return new QueryInstancesClient(nacosClient, nacosApiVersion, serviceName, namespaceId, groupName,
+                                        clusterName, healthyOnly, app);
+    }
+
+    private final WebClient webClient;
+
+    private final String pathForQuery;
+
+    QueryInstancesClient(NacosClient nacosClient, String nacosApiVersion, String serviceName,
+                         @Nullable String namespaceId, @Nullable String groupName,
+                         @Nullable String clusterName, @Nullable Boolean healthyOnly, @Nullable String app) {
+        webClient = nacosClient.nacosWebClient();
+
+        final StringBuilder pathBuilder = new StringBuilder("/")
+                .append(nacosApiVersion)
+                .append("/ns/instance/list?");
+        final QueryParams params = NacosClientUtil
+                .queryParams(namespaceId, groupName, requireNonNull(serviceName, "serviceName"),
+                        clusterName, healthyOnly, app, null, null, null);
+        pathBuilder.append(params.toQueryString());
+
+        pathForQuery = pathBuilder.toString();
+    }
+
+    CompletableFuture<List<Endpoint>> endpoints() {
+        return queryInstances()
+                .thenApply(response -> response.data.hosts.stream()
```

Comment:
Question) Just to make sure, even if `code != 0` (which I assume means successful) will `data` and `hosts` always exist?
I'm asking because there doesn't seem to defence against `null` check

Reply:
@jrhee17 

I added null check for data and data.hosts! Thanks.